File: db-objects/src/main/scala/com/simplesys/container/scala/OrdSource.scala
package com.simplesys.container.scala

import java.sql.Timestamp
import java.time.LocalDateTime

import com.simplesys.container.Helper
import com.simplesys.container.java.JOrdSource
import oracle.jdbc.OracleBlob
import oracle.sql.NUMBER

object OrdSource {

    implicit def JOrdSource2SOrdSource(ordSource: JOrdSource): OrdSource = new OrdSource {
        override val srcName: Option[String] = Option(Helper.asString(ordSource.srcName))
        override val srcLocation: Option[String] = Option(Helper.asString(ordSource.srcLocation))
        override val updateTime: Option[LocalDateTime] = Option(ordSource.updateTime).map(_.toLocalDateTime)
        override val local: Option[BigDecimal] = Option(ordSource.local).map(Helper.asBigDecimal(_))
        override val srcType: Option[String] = Option(Helper.asString(ordSource.srcType))
        override val localData: Option[OracleBlob] = Option(ordSource.localData)
    }

    implicit def SOrdSource2JOrdSource(ordSource: OrdSource): JOrdSource = {
        val res = new JOrdSource {}
        ordSource.srcName.foreach(srcName ⇒ res.srcName = Helper.asCHAR(srcName))
        ordSource.srcLocation.foreach(srcLocation => res.srcLocation = Helper.asCHAR(srcLocation))
        ordSource.updateTime.foreach(localDateTime ⇒ res.updateTime = Timestamp.valueOf(localDateTime))
        ordSource.local.foreach(bigDeecimal ⇒ res.local = new NUMBER(bigDeecimal.bigDecimal))
        ordSource.srcType.foreach(srcType ⇒ res.srcType = Helper.asCHAR(srcType))
        ordSource.localData.foreach(res.localData = _)
        res
    }
}

trait OrdSource {
    val localData: Option[OracleBlob]
    val srcType: Option[String]
    val srcLocation: Option[String]
    val srcName: Option[String]
    val updateTime: Option[LocalDateTime]
    val local: Option[BigDecimal]

    override def toString: String = {
        val res = new StringBuilder
        res append s"localData: $localData ,"
        res append s"srcType: $srcType ,"
        res append s"srcLocation: $srcLocation ,"
        res append s"srcName: $srcName, "
        res append s"updateTime: $updateTime ,"
        res append s"local: $local"
        res.toString()
    }
}

File: db-objects/src/main/java/com/simplesys/container/java/OrdSource.java
package com.simplesys.container.java;

import com.simplesys.container.Helper;
import oracle.jdbc.OracleBlob;
import oracle.jdbc.OracleData;
import oracle.jdbc.OracleDataFactory;
import oracle.jdbc.internal.OracleStruct;
import oracle.sql.NUMBER;

import java.sql.Connection;
import java.sql.SQLException;
import java.sql.Timestamp;

public class OrdSource implements OracleData, OracleDataFactory {

  public OracleBlob localData;
  public String srcType;
  public String srcLocation;
  public String srcName;
  public Timestamp updateTime;
  public NUMBER local;

  static final OrdSource _ordSource = new OrdSource();

  public static OracleDataFactory getOracleDataFactory() {
    return _ordSource;
  }

  public OrdSource() {
  }

  @Override
  public String toString() {
    try {
      return "localData: BLOB" + " ,srcType: " + srcType + " ,srcLocation: " + srcLocation + " ,srcName: " + srcName + " ,updateTime: " + updateTime.toString() + " ,local: " + Helper.asBigDecimal(local);
    } catch (SQLException e) {
      return e.getMessage();
    }
  }

  public OrdSource(OracleBlob localData,
                   String srcType,
                   String srcLocation,
                   String srcName,
                   Timestamp updateTime,
                   NUMBER local) {
    this.localData = localData;
    this.srcType = srcType;
    this.srcLocation = srcLocation;
    this.srcName = srcName;
    this.updateTime = updateTime;
    this.local = local;
  }

  @Override
  public Object toJDBCObject(Connection connection) throws SQLException {
    Object[] attributes = {localData, srcType, srcLocation, srcName, updateTime, local};
    return connection.createStruct("ORDSYS.ORDSOURCE", attributes);
  }

  @Override
  public OracleData create(Object o, int i) throws SQLException {
    if (o == null)
      return null;

    Object[] attributes = ((OracleStruct) o).getOracleAttributes();
    return new OrdSource(
            (OracleBlob) attributes[0],
            (String) attributes[1],
            (String) attributes[2],
            (String) attributes[3],
            (Timestamp) attributes[4],
            (NUMBER) attributes[5]);
  }
}
